Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java

```diff
@@ -1004,42 +1066,117 @@ public RandomCutForest build() {
+            state.setThreadPoolSize(forest.threadPoolSize);
+            state.setStoreSequenceIndexesEnabled(forest.storeSequenceIndexesEnabled);
+            state.setEntreesSeen(forest.getTotalUpdates());
+            state.setCompactEnabled(forest.compactEnabled);
+            state.setSaveTreeData(forest.saveTreeData);
+
+            if (forest.compactEnabled) {
+                PointStoreCoordinator pointStoreCoordinator = (PointStoreCoordinator) forest.updateCoordinator;
+                PointStoreDoubleState pointStoreState = new PointStoreDoubleMapper()
+                        .toState(pointStoreCoordinator.getStore());
+                state.setPointStoreDoubleState(pointStoreState);
+
+                List<CompactSamplerState> samplerStates = new ArrayList<>();
+                List<CompactRandomCutTreeState> treeStates = null;
+                if (saveTreeState) {
+                    treeStates = new ArrayList<>();
+                }
+
+                CompactSamplerMapper samplerMapper = new CompactSamplerMapper();
+                CompactRandomCutTreeMapper treeMapper = new CompactRandomCutTreeMapper();
+
+                for (IComponentModel<?> model : forest.componentModels) {
+                    SamplerPlusTree<Integer> samplerPlusTree = (SamplerPlusTree<Integer>) model;
+                    CompactSampler sampler = (CompactSampler) samplerPlusTree.getSampler();
+                    samplerStates.add(samplerMapper.toState(sampler));
+                    if (treeStates != null) {
+                        CompactRandomCutTreeDouble tree = (CompactRandomCutTreeDouble) samplerPlusTree.getTree();
+                        treeStates.add(treeMapper.toState(tree));
+                    }
+                }
+                state.setCompactSamplerStates(samplerStates);
+                state.setCompactRandomCutTreeStates(treeStates);
+            } else {
+                ArraySamplersToCompactStateConverter converter = new ArraySamplersToCompactStateConverter(
+                        forest.storeSequenceIndexesEnabled, forest.dimensions, forest.sampleSize);
+
+                for (IComponentModel<?> model : forest.componentModels) {
+                    SamplerPlusTree<double[]> samplerPlusTree = (SamplerPlusTree<double[]>) model;
+                    SimpleStreamSampler<double[]> sampler = (SimpleStreamSampler<double[]>) samplerPlusTree
+                            .getSampler();
+                    converter.addSampler(sampler);
+                }
+
+                state.setPointStoreDoubleState(converter.getPointStoreDoubleState());
+                state.setCompactSamplerStates(converter.getCompactSamplerStates());
+            }
+
+            return state;
+        }
+
+        @Override
+        public RandomCutForest toModel(RandomCutForestState state, long seed) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public RandomCutForestState getForestState() {
```

Comment:
question. is this a duplicate with Forest.StateMapper line 1093?

Reply:
Yes, eventually the new state mapper could should replace the getForestState(). Currently the state mapper code is incomplete, and getForestState() is being used in the serde tests.